src/backend/gslice.c

```diff
@@ -0,0 +1,256 @@
+                return;
+        }
+    }
+}
+
+static void sliceStructs_Replace(SymInfo *sia, elem *e)
+{
+    while (1)
+    {
+        switch (e->Eoper)
+        {
+            case OPvar:
+            {
+                Symbol *s = e->EV.sp.Vsym;
+                SYMIDX si = s->Ssymnum;
+                //printf("e: %d %d\n", si, sia[si].canSlice);
+                //elem_print(e);
+                if (si >= 0 && sia[si].canSlice)
+                {
+                    if (e->Eoffset == 0)
+                    {
+                        type_free(s->Stype);
+                        s->Stype = type_fake(e->Ety);
+                    }
+                    else
+                    {
+                        Symbol *s1 = globsym.tab[sia[si].si0 + 1];
+                        type_free(s1->Stype);
+                        s1->Stype = type_fake(e->Ety);
+                        e->EV.sp.Vsym = s1;
+                        e->Eoffset = 0;
+                        //printf("replaced with:\n");
+                        //elem_print(e);
+                    }
+                }
+                return;
+            }
+
+            default:
+                if (OTunary(e->Eoper))
+                {
+                    e = e->E1;
+                    break;
+                }
+                if (OTbinary(e->Eoper))
+                {
+                    sliceStructs_Replace(sia, e->E2);
+                    e = e->E1;
+                    break;
+                }
+                return;
+        }
+    }
+}
+
+void sliceStructs()
+{
+    if (debugc) printf("sliceStructs()\n");
+    size_t sia_length = globsym.top;
+    SymInfo *sia = (SymInfo *)malloc(3 * sia_length * sizeof(SymInfo));
+    assert(sia);
+    SymInfo *sia2 = sia + sia_length;
+
+    bool anySlice = false;
+    for (int si = 0; si < globsym.top; si++)
+    {
+        Symbol *s = globsym.tab[si];
+        //printf("slice1: %s\n", s->Sident);
+
+        if ((s->Sflags & (GTregcand | SFLunambig)) != (GTregcand | SFLunambig))
+        {
+            sia[si].canSlice = false;
+            continue;
+        }
+
+        targ_size_t sz = type_size(s->Stype);
+        if (sz != 2 * REGSIZE ||
+            tyfv(s->Stype->Tty) || tybasic(s->Stype->Tty) == TYhptr)    // because there is no TYseg
+        {
+            sia[si].canSlice = false;
+            continue;
+        }
+
+        switch (s->Sclass)
+        {
+            case SCfastpar:
+            case SCregister:
+            case SCauto:
+            case SCshadowreg:
+            case SCparameter:
+                anySlice = true;
+                sia[si].canSlice = true;
+                break;
+
+            case SCstack:
+            case SCpseudo:
+            case SCstatic:
+            case SCbprel:
+                sia[si].canSlice = false;
+                break;
+
+            default:
+                symbol_print(s);
+                assert(0);
+        }
+    }
+
+    if (!anySlice)
+        goto Ldone;
```

Comment:
I made it a struct so I can add more things to it later. It's very unlikely to get large, as it's the number of local symbols in a function, so size isn't a problem.